ssl/ssl_sess.c

```diff
@@ -1081,58 +1102,76 @@ static void SSL_SESSION_list_remove(SSL_CTX *ctx, SSL_SESSION *s)
     s->prev = s->next = NULL;
 }
 
+/* session cache is already locked by caller */
 static void SSL_SESSION_list_add(SSL_CTX *ctx, SSL_SESSION *s)
 {
-    if ((s->next != NULL) && (s->prev != NULL))
-        SSL_SESSION_list_remove(ctx, s);
+    SSL_SESSION_CACHE *session_cache = ctx->session_cache;
 
-    if (ctx->session_cache_head == NULL) {
-        ctx->session_cache_head = s;
-        ctx->session_cache_tail = s;
-        s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
-        s->next = (SSL_SESSION *)&(ctx->session_cache_tail);
+    if (session_cache == NULL)
+        return;
+
+    SSL_SESSION_list_remove(ctx, s);
+
+    if (session_cache->head == NULL) {
+        session_cache->head = s;
+        session_cache->tail = s;
+        s->prev = (SSL_SESSION *)&(session_cache->head);
+        s->next = (SSL_SESSION *)&(session_cache->tail);
     } else {
-        s->next = ctx->session_cache_head;
+        s->next = session_cache->head;
         s->next->prev = s;
-        s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
-        ctx->session_cache_head = s;
+        s->prev = (SSL_SESSION *)&(session_cache->head);
+        session_cache->head = s;
     }
 }
 
 void SSL_CTX_sess_set_new_cb(SSL_CTX *ctx,
                              int (*cb) (struct ssl_st *ssl, SSL_SESSION *sess))
 {
-    ctx->new_session_cb = cb;
+    CRYPTO_THREAD_write_lock(ctx->lock);
+    CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+    ctx->session_cache->new_session_cb = cb;
+    CRYPTO_THREAD_unlock(ctx->session_cache->lock);
+    CRYPTO_THREAD_unlock(ctx->lock);
```

Comment:
What I'm trying to say is that this seems like case of excessive locking. Or on certain level meaningless locking. Don't callbacks kind of define how SSL_SESSION operates and should be set prior it becomes appropriate to share it? Because in such case, locking is not necessary...